### ISSUE TITLE
d3js visualization

### ISSUE THREAD

**Issue by Lompik** [melpa/melpa#3818]
Hi

In melpa.js I saw

    // TODO D3 visualisation for deps

I've actually made one https://github.com/Lompik/emacs-melpa-network based on the json served on melpa. An automatic build is doable with nodejs but I don't have this yet. 

Basically, I'm looking to see if that TODO is still pending/current and if I help complete it in some way.

Regards.

**Comment by purcell** [comment]
That's very neat! I hadn't really thought of trying to display all the packages on a single chart -- what I had in mind was to be able to show a tree of all directly and indirectly dependent packages for any specific package, so on the [page for `s`](http://melpa.org/#/s), for example, I'd love to see a graph showing not just the "Needed by" list of packages, but also the packages which depend on those.

**Comment by Lompik** [comment]
You mean like :
https://fab974.org/blog/melpa-d3js/spackagedeps.svg

the `s` package is probably one of the most filled in term of dependency with 230 packages.

**Comment by purcell** [comment]
Haven't forgotten about this - just very busy, and finding it hard to tackle bigger things. One of the things I'm wondering is how to integrate the d3 diagrams with the rest of the mithril.js code.

**Comment by Lompik** [comment]
I'm not an expert but it seems we can can call methods on DOM element via the config attribute .. and have the d3 in the mehods. https://lhorie.github.io/mithril/integration.html and https://lhorie.github.io/mithril/mithril.html#the-config-attribute.

AFAIU you want to have the d3 code inside https://github.com/melpa/melpa/blob/master/html/js/melpa.js#L108 and process the json data from there, on the client side. I was more thinking about generating the 3000+ svgs in the `Makefile`. 
I guess the question I have for you is: how much interactiveness are you planning for the graphs ? 

There are plenty of discrepencies in json data (wrong depencies names, version, etc) that comes from the providers. Maybe tring to fix could be a first step.

**Comment by purcell** [comment]
> AFAIU you want to have the d3 code inside https://github.com/melpa/melpa/blob/master/html/js/melpa.js#L108 and process the json data from there, on the client side. I was more thinking about generating the 3000+ svgs in the Makefile.

Yeah, I guess so, because the json data is already available in the client, and we don't pre-process much data on the server. Issue #3728 relates to switching to generating most of the website content on the server, but that's non-trivial esssentially due to cache invalidation. :-)

> I guess the question I have for you is: how much interactiveness are you planning for the graphs ?

Probably just links through to the corresponding MELPA package pages.

> There are plenty of discrepencies in json data (wrong depencies names, version, etc) that comes from the providers. Maybe trying to fix could be a first step.

Oh! Do you mean that you found broken packages? If so, I'd definitely want to get those fixed.
